ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateReader.scala

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import akka.NotUsed
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueConsumption}
+import com.daml.ledger.participant.state.v1._
+import com.digitalasset.daml.lf.data.Time
+import com.digitalasset.ledger.api.health.{HealthStatus, Healthy}
+
+class KeyValueParticipantStateReader(reader: LedgerReader)(implicit materializer: Materializer)
+    extends ReadService {
+  override def getLedgerInitialConditions(): Source[LedgerInitialConditions, NotUsed] =
+    Source.single(createLedgerInitialConditions())
+
+  override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] =
+    reader
+      .events(toReaderOffset(beginAfter))
+      .flatMapConcat { record =>
+        val updates = Envelope.open(record.envelope) match {
+          case Right(Envelope.LogEntryMessage(logEntry)) =>
+            KeyValueConsumption
+              .logEntryToUpdate(record.entryId, logEntry)
+              .zipWithIndex
+              .map {
+                case (entry, index) =>
+                  (toReturnedOffset(index, record.offset), entry)
+              }
+          case _ => Seq.empty
```

Comment:
Should we just throw here? If we get other data than log entries we have a bug.

Reply:
Changed it to throw in case of an invalid log entry.